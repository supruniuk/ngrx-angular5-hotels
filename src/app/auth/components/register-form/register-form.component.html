<mat-card class="register-form">
  <mat-card-title>Register</mat-card-title>
  <mat-card-content>
    <form [formGroup]="form" (ngSubmit)="submit()" fxLayout="column">

      <mat-form-field>
        <mat-select placeholder="Account type" required formControlName="authType">
          <mat-option *ngFor="let type of ['customer', 'hotel']" [value]="type">
            {{ type }}
          </mat-option>
        </mat-select>
      </mat-form-field>

      <mat-form-field>
        <input type="text" matInput required placeholder="User name" formControlName="username">
      </mat-form-field>    

      <mat-form-field>
        <input type="text" matInput required placeholder="Email" formControlName="email">
      </mat-form-field>

      <mat-form-field>
        <input type="password" matInput required placeholder="Password" formControlName="password">
      </mat-form-field>

      <p *ngIf="errorMessage" class="register-error">
        {{ errorMessage }}
      </p>          
    
      <div class="register-action-buttons" fxLayoutGap="32px">
        <button type="button" mat-button (click)="cancel()">Cancel</button> 
        <button mat-raised-button color="primary" type="submit">Create account</button> 
      </div>

    </form>
  </mat-card-content>
</mat-card>