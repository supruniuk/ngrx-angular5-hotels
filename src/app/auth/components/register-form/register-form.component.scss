// :host {
//   display: flex;
//   justify-content: center;
//   // flex-direction: row;
//   // height: 100%;
//   // margin: 72px 0;
// }

.register-form {
  min-width: 300px;
}

.register-error {
  padding: 16px;
  color: white;
  background-color: red;
}

.register-action-buttons {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}