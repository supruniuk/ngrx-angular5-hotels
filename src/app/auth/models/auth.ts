import { User } from './user';


export interface Authenticate {
  email: string;
  password: string;
}

export interface SignUp extends Authenticate {
  authType: string;
  name: string;
}

export interface AuthResponse {
  user: User;
  token: string;
}

export const CUSTOMER_TYPE = 'customer';
export const HOTEL_TYPE = 'hotel';

export class Auth {
  constructor(
    public token: string,
    public type: string
  ) { }
}

// export class CustomerAuth extends Auth {
//   readonly type = CUSTOMER_TYPE;
// }

// export class HotelAuth extends Auth {
//   readonly type = HOTEL_TYPE;
// }
