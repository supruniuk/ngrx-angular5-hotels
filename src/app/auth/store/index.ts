export * from './actions/auth';
export * from './reducers';
export * from './effects/auth.effects';
export * from './selectors';
