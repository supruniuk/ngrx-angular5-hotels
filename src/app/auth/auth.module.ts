import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { ReactiveFormsModule } from '@angular/forms';
import { FlexLayoutModule } from '@angular/flex-layout';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { Routes } from '@angular/router';

import { MaterialModule } from '../material';
import { LoginPageComponent } from './containers/login-page/login-page.component';
import { RegisterPageComponent } from './containers/register-page/register-page.component';
import { LoginFormComponent } from './components/login-form/login-form.component';
import { RegisterFormComponent } from './components/register-form/register-form.component';

import { AuthInterceptor, AuthService, AuthGuard, RoleGuard } from './services';
import { AuthEffects, reducers, metaReducers } from './store';


export const COMPONENTS = [
  LoginPageComponent,
  LoginFormComponent,
  RegisterPageComponent,
  RegisterFormComponent,
];

export const ROUTES: Routes = [
  { path: 'login', component: LoginPageComponent },
  { path: 'register', component: RegisterPageComponent }
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    FlexLayoutModule
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class AuthModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: RootAuthModule,
      providers: [
        AuthService,
        AuthGuard,
        RoleGuard,
        {
          provide: HTTP_INTERCEPTORS,
          useClass: AuthInterceptor,
          multi: true
        },
      ],
    };
  }
}

@NgModule({
  imports: [
    AuthModule,
    RouterModule.forChild(ROUTES),
    StoreModule.forFeature('auth', reducers, { metaReducers }),
    EffectsModule.forFeature([AuthEffects]),
  ],
})
export class RootAuthModule { }
