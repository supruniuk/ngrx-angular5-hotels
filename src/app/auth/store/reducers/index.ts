import { localStorageSync } from 'ngrx-store-localstorage';
import {
  Action,
  ActionReducer,
  createFeatureSelector,
  createSelector,
  MetaReducer
} from '@ngrx/store';
import { RootState } from '../../../core/store/reducers';
import * as fromStatus from './auth-status.reducers';
import * as fromLoginPage from './login-page.reducers';
import * as fromRegisterPage from './register-page.reducers';


export interface AuthState {
  status: fromStatus.StatusState;
  loginPage: fromLoginPage.LoginState;
  registerPage: fromRegisterPage.RegisterState;
}

export interface RootAuthState extends RootState {
  auth: AuthState;
}

export const reducers = {
  status: fromStatus.reducer,
  loginPage: fromLoginPage.reducer,
  registerPage: fromRegisterPage.reducer,
};

// Setting up ngrx-store-localstorage reducer
// to keep authentication info in the LocalStorage
export function localStorageSyncReducer(reducer: ActionReducer<any>): ActionReducer<any> {
  return localStorageSync({ keys: ['status'], rehydrate: true })(reducer);
}

export const metaReducers: Array<MetaReducer<any, any>> = [localStorageSyncReducer];

export const getAuthState = createFeatureSelector<AuthState>('auth');
