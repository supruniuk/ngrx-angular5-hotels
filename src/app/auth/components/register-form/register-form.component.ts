import { Component, Input, Output, EventEmitter } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
import { SignUp } from '../../models/auth';

@Component({
  selector: 'hot-register-form',
  templateUrl: './register-form.component.html',
  styleUrls: ['./register-form.component.scss'],
})
export class RegisterFormComponent {
  @Input()
  set pending(isPending: boolean) {
    if (isPending) {
      this.form.disable();
    } else {
      this.form.enable();
    }
  }

  @Input() errorMessage: string | null;

  @Output() onSubmit = new EventEmitter<SignUp>();

  @Output() onCancel = new EventEmitter();

  form: FormGroup = new FormGroup({
    authType: new FormControl(''),
    name: new FormControl(''),
    email: new FormControl(''),
    password: new FormControl(''),
  });

  constructor() {}

  submit() {
    if (this.form.valid) {
      this.onSubmit.emit(this.form.value);
    }
  }

  cancel() {
    this.onCancel.emit();
  }
}
