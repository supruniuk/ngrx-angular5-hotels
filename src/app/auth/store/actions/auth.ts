import { Action } from '@ngrx/store';
import { User } from '../../models/user';
import { Authenticate, SignUp, Auth } from '../../models/auth';

export enum AuthActionTypes {
  login = '[Auth] Login',
  logout = '[Auth] Logout',
  loginSuccess = '[Auth] Login Success',
  loginFail = '[Auth] Login Failure',
  loginRedirect = '[Auth] Login Redirect',
  customerRedirect = '[Auth] Customer Redirect',
  hotelRedirect = '[Auth] Hotel Redirect',
  register = '[Auth] Register',
  registerSuccess = '[Auth] Register Success',
  registerFail = '[Auth] Register Failure',
  registerRedirect = '[Auth] Register Redirect',
}

export class Login implements Action {
  readonly type = AuthActionTypes.login;

  constructor(public payload: Authenticate) {}
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.loginSuccess;

  constructor(public payload: { user: User, auth: Auth }) {}
}

export class LoginFail implements Action {
  readonly type = AuthActionTypes.loginFail;

  constructor(public payload: any) {}
}

export class LoginRedirect implements Action {
  readonly type = AuthActionTypes.loginRedirect;
}

export class CustomerRedirect implements Action {
  readonly type = AuthActionTypes.customerRedirect;
}

export class HotelRedirect implements Action {
  readonly type = AuthActionTypes.hotelRedirect;
}

export class Register implements Action {
  readonly type = AuthActionTypes.register;

  constructor(public payload: SignUp) {}
}

export class RegisterSuccess implements Action {
  readonly type = AuthActionTypes.registerSuccess;

  constructor(public payload: { user: User, auth: Auth }) {}
}

export class RegisterFail implements Action {
  readonly type = AuthActionTypes.registerFail;

  constructor(public payload: any) {}
}

export class RegisterRedirect implements Action {
  readonly type = AuthActionTypes.registerRedirect;

  constructor(public payload: string) {}
}

export class Logout implements Action {
  readonly type = AuthActionTypes.logout;
}

export type AuthActions = Login
  | LoginSuccess
  | LoginFail
  | LoginRedirect
  | CustomerRedirect
  | HotelRedirect
  | Register
  | RegisterSuccess
  | RegisterFail
  | RegisterRedirect
  | Logout;
