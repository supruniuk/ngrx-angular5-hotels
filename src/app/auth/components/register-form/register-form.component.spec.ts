import { TestBed, ComponentFixture } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { StoreModule, Store, combineReducers } from '@ngrx/store';
import { RegisterFormComponent } from './register-form.component';
import * as Auth from '../../actions/auth';
import * as fromAuth from '../../reducers';
import { ReactiveFormsModule } from '@angular/forms';

describe('Register Page', () => {
  let fixture: ComponentFixture<RegisterFormComponent>;
  let instance: RegisterFormComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [ReactiveFormsModule],
      declarations: [RegisterFormComponent],
      schemas: [NO_ERRORS_SCHEMA],
    });

    fixture = TestBed.createComponent(RegisterFormComponent);
    instance = fixture.componentInstance;
  });

  it('should compile', () => {
    fixture.detectChanges();

    /**
     * The login form is a presentational component, as it
     * only derives its state from inputs and communicates
     * externally through outputs. We can use snapshot
     * tests to validate the presentation state of this component
     * by changing its inputs and snapshotting the generated
     * HTML.
     *
     * We can also use this as a validation tool against changes
     * to the component's template against the currently stored
     * snapshot.
     */
    expect(fixture).toMatchSnapshot();
  });

  it('should disable the form if pending', () => {
    instance.pending = true;

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should display an error message if provided', () => {
    instance.errorMessage = 'Invalid credentials';

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should emit an event if the form is valid when submitted', () => {
    const credentials = {
      name: 'user',
      email: 'user@mail.com',
      password: 'pass',
    };
    instance.form.setValue(credentials);

    spyOn(instance.onSubmit, 'emit');
    instance.submit();

    expect(instance.onSubmit.emit).toHaveBeenCalledWith(credentials);
  });
});
