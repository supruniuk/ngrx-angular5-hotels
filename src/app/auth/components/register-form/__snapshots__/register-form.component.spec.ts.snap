// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Page should compile 1`] = `
<hot-register-form
  form={[Function FormGroup]}
  onCancel={[Function EventEmitter]}
  onSubmit={[Function EventEmitter]}
>
  <mat-card>
    
    
    <mat-card-title>
      Register
    </mat-card-title>
    
    
    <mat-card-content>
      
      
      <form
        class="ng-untouched ng-pristine ng-invalid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="name"
              matInput=""
              ng-reflect-name="name"
              ng-reflect-required=""
              placeholder="Name"
              required=""
              type="text"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="email"
              matInput=""
              ng-reflect-name="email"
              ng-reflect-required=""
              placeholder="Email"
              required=""
              type="text"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="password"
              matInput=""
              ng-reflect-name="password"
              ng-reflect-required=""
              placeholder="Password"
              required=""
              type="password"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        
                  
      
        
        <p
          class="registerButtons"
        >
          
          
          <button
            mat-button=""
            type="button"
          >
            Cancel
          </button>
           
          
          <button
            color="primary"
            mat-raised-button=""
            type="submit"
          >
            Create account
          </button>
            
        
        </p>
        
  
      
      </form>
      
    
    </mat-card-content>
    
  
  </mat-card>
  
  
  
  
  
  
</hot-register-form>
`;

exports[`Register Page should disable the form if pending 1`] = `
<hot-register-form
  form={[Function FormGroup]}
  onCancel={[Function EventEmitter]}
  onSubmit={[Function EventEmitter]}
>
  <mat-card>
    
    
    <mat-card-title>
      Register
    </mat-card-title>
    
    
    <mat-card-content>
      
      
      <form
        class="ng-untouched ng-pristine"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine"
              disabled=""
              formcontrolname="name"
              matInput=""
              ng-reflect-name="name"
              ng-reflect-required=""
              placeholder="Name"
              required=""
              type="text"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine"
              disabled=""
              formcontrolname="email"
              matInput=""
              ng-reflect-name="email"
              ng-reflect-required=""
              placeholder="Email"
              required=""
              type="text"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine"
              disabled=""
              formcontrolname="password"
              matInput=""
              ng-reflect-name="password"
              ng-reflect-required=""
              placeholder="Password"
              required=""
              type="password"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        
                  
      
        
        <p
          class="registerButtons"
        >
          
          
          <button
            mat-button=""
            type="button"
          >
            Cancel
          </button>
           
          
          <button
            color="primary"
            mat-raised-button=""
            type="submit"
          >
            Create account
          </button>
            
        
        </p>
        
  
      
      </form>
      
    
    </mat-card-content>
    
  
  </mat-card>
  
  
  
  
  
  
</hot-register-form>
`;

exports[`Register Page should display an error message if provided 1`] = `
<hot-register-form
  errorMessage={[Function String]}
  form={[Function FormGroup]}
  onCancel={[Function EventEmitter]}
  onSubmit={[Function EventEmitter]}
>
  <mat-card>
    
    
    <mat-card-title>
      Register
    </mat-card-title>
    
    
    <mat-card-content>
      
      
      <form
        class="ng-untouched ng-pristine ng-invalid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="name"
              matInput=""
              ng-reflect-name="name"
              ng-reflect-required=""
              placeholder="Name"
              required=""
              type="text"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="email"
              matInput=""
              ng-reflect-name="email"
              ng-reflect-required=""
              placeholder="Email"
              required=""
              type="text"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        <p>
          
          
          <mat-form-field>
            
            
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="password"
              matInput=""
              ng-reflect-name="password"
              ng-reflect-required=""
              placeholder="Password"
              required=""
              type="password"
            />
            
          
          </mat-form-field>
          
        
        </p>
        
  
        
        
        <p
          class="registerError"
        >
          
          Invalid credentials
        
        </p>
                  
      
        
        <p
          class="registerButtons"
        >
          
          
          <button
            mat-button=""
            type="button"
          >
            Cancel
          </button>
           
          
          <button
            color="primary"
            mat-raised-button=""
            type="submit"
          >
            Create account
          </button>
            
        
        </p>
        
  
      
      </form>
      
    
    </mat-card-content>
    
  
  </mat-card>
  
  
  
  
  
  
</hot-register-form>
`;
